### ISSUE TITLE
Pattern Matching with Tensors

### ISSUE THREAD

**Issue by plotkiny** [pytorch/pytorch#72737]
### 🚀 The feature, motivation and pitch

Taking a nod from sequence alignment algorithms in bioinformatics, given two tensors, a _template_ and _pattern_, it would be a nice feature in PyTorch to be able to find the indices of the pattern in the template. 

The pattern matching algorithm can be exact (needs to match the identical sequence in the pattern) or approximate (given the maximum number of mismatches). I provide a few examples below.

Given a template tensor:

`template = torch.tensor([101.,   124.,   979.,   430.,  1073.,   146.,   111.,   650.,   983., 430., 25484.,   983.,  2731.,   111.])`

Given the following pattern tensor:

`pattern1= torch.tensor([ 430.,  1073.,   146.,   111.,   650.])`

**Find the indices of the exact pattern in the template:**

`start_indices = torch.match(pattern1, template, mismatch_cnt=0, axis=0)` 
`start_indices = torch.tensor([3])`

**Find the indices of the approximate pattern in the template:**

Given the following pattern tensor with 2 mismatches (indices 2 and 4 do not match the sequence in the template):

`pattern2 = torch.tensor([ 111.,   650.,   101.,  430.,  430., 983.])`

`start_indices = torch.match(pattern2,  template, max_mismatch=2, axis=0)` 
`start_indices = torch.tensor([6])`

The axis dimension is provided for searching tensors in other dimensions.

### Alternatives

I've attached a file with the python code for the approximate Boyer Moore pattern matching algorithm using two lists as inputs. 

[approximate_match_lists.txt](https://github.com/pytorch/pytorch/files/8051323/approximate_match_lists.txt)

**Comment by soulitzer** [comment]
Interesting idea. Do you know if there are any other libraries have this already?

**Comment by plotkiny** [comment]
Thanks. I've searched though surprisingly, to the best of my knowledge, I did not find any libraries that could work with tensors.

**Comment by Neloy262** [comment]
@plotkiny what might be the use case for this?
